### PR TITLE
Add option to toggle console window

### LINKED ISSUE #17889: Add an option to MS Windows Binary to launch without console window
**Godot version:**
3.0.2

**OS/device including version:**
Microsoft Windows 10

It would be nice to have an option to launch the engine without an attached console window on Windows OS. 

Just little thing, but it makes switching between Godot and other applications one click more, which adds up when switching often. It also gets confusing when having more then one Godot instance open.

### PR DESCRIPTION
**Pull request by lukad**
This adds `Editor -> Toggle System Console` to the menu bar on Windows and behaves like Blender's "Toggle System Console".
The setting is not persisted but it still makes working on windows a lot easier.

*Bugsquad edit:* Fixes #17889.

### CODE REVIEW

**Comment by SeleckyErik** [comment]
It might be even useful to store this setting in the editor and/or project settings. That way, if preferred workspace setting is to have the console hidden, you don't have to hide it every time you open a project. Adding it to project settings too would have a benefit of enabling overriding this default on a per-project basis - although I am not sure how useful that would be.

**Comment by lukad** [comment]
I think it would be best to have this in the editor settings. If there really is a need to override this for some projects it can be implemented later.
I'll look into this (don't have much experience with the godot code).

File: core/os/os.h [hunk 1]
```diff
@@ -110,10 +111,11 @@ class OS {
 		bool layered_splash;
 		bool layered;
 		float get_aspect() const { return (float)width / (float)height; }
-		VideoMode(int p_width = 1024, int p_height = 600, bool p_fullscreen = false, bool p_resizable = true, bool p_borderless_window = false, bool p_maximized = false, bool p_always_on_top = false, bool p_use_vsync = false) {
+		VideoMode(int p_width = 1024, int p_height = 600, bool p_fullscreen = false, bool p_console = true, bool p_resizable = true, bool p_borderless_window = false, bool p_maximized = false, bool p_always_on_top = false, bool p_use_vsync = false) {
```
Comment (reduz) [inline core/os/os.h]: I don't think it's worth putting this here, just let it on by default on debug I guess.

File: editor/editor_node.cpp [hunk 2]
```diff
@@ -2261,6 +2261,11 @@ void EditorNode::_menu_option_confirm(int p_option, bool p_confirmed) {
 
 			OS::get_singleton()->set_window_fullscreen(!OS::get_singleton()->is_window_fullscreen());
 
+		} break;
+		case SETTINGS_TOGGLE_CONSOLE: {
+
+			OS::get_singleton()->set_console_visible(!OS::get_singleton()->is_console_visible());
+
```
Comment (reduz) [inline editor/editor_node.cpp]: you could also save this as a default value using EDITOR_DEF, EDITOR_GET

**Review by reduz — DISMISSED** [review]

**Comment by rogeriodec** [comment]
Found a workaround for Windows, while this is not implemented: [RBTray](http://rbtray.sourceforge.net/).

_RBTray is a small Windows program that runs in the background and allows almost any window to be minimized to the system tray by right-clicking its minimize button._

**Comment by Calinou** [comment]
To follow up @rogeriodec's comment, an up-to-date RBTray fork that works better with Windows 10 can be found here: https://github.com/benbuck/rbtray

**Comment by rogeriodec** [comment]
For those who want something even more automated and do not mind having the console hidden, using [Autohotkey](https://www.autohotkey.com/), you can use [this simple script](https://drive.google.com/open?id=1p-cnoUZ8SwBvQoudxticNS7voexeOCPm) I made. 
It gets active in the windows taskbar and detects when Godot is activated.

**Comment by MumChamber** [comment]
Hi @lukad, any chance of updating wiht @reduz's changes?

**Comment by akien-mga** [comment]
I force pushed an amend with the changes requested by @reduz plus some more fixes.

Tested on Windows 10, seems to work fine.

I put some more details in the commit log:

This is an editor setting and its value can also be toggled
using an entry in the Editor toolbar. The console will still
appear briefly when starting the project manager or editor,
as it's still compiled as console application.

Does not impact exported games, which will still run without
console in release and with console in debug mode.

A project setting or export option could be added to disable
it in debug mode if there's demand for it, but that would
greatly reduce the usefulness of debug builds if Windows users
can no longer report error and crash messages.

**Comment by akien-mga** [comment]
For now the setting is opt-in, so the console will be shown by default (no change). I could switch it to opt-out (console hidden by default), though it would make it harder to notice and could likely reduce the quality of bug reports we'd get from Windows users.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**